app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/gainorlossafterreliefs/GlarCalculatorEligibilityUtil.scala

```diff
@@ -0,0 +1,151 @@
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendController
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class GlarCalculatorEligibilityUtil @Inject()(
+                                               val authenticatedAction: AuthenticatedAction,
+                                               val sessionDataAction: SessionDataAction,
+                                               val sessionStore: SessionStore,
+                                               val errorHandler: ErrorHandler,
+                                               returnsService: ReturnsService,
+                                               cc: MessagesControllerComponents,
+                                               servicesConfig: ServicesConfig
+                                             )(implicit ec: ExecutionContext)
+  extends FrontendController(cc)
+    with WithAuthAndSessionDataAction
+    with SessionUpdates
+    with Logging
+    with StartingToAmendToFillingOutReturnBehaviour {
+
+  val glarCalculatorFlag = servicesConfig.getBoolean("glar-calculator.enabled")
+
+  def withValuesForCalculatorWhereEligible(fillingOutReturn: FillingOutReturn)(
+    f: Option[CalculatedGlarBreakdown] => Future[Result]
+  )(implicit request: RequestWithSessionData[_]): Future[Result] = {
+
+    def isEligible(
+                    fillingOutReturn: FillingOutReturn,
+                    completeSingleDisposalTriageAnswers: CompleteSingleDisposalTriageAnswers,
+                    reliefDetailsAnswers: CompleteReliefDetailsAnswers
+                  )(f: Boolean => Future[Result]): Future[Result] =
+      if (!glarCalculatorFlag) {
+        f(false)
+      } else {
+        val eligibleAssetType              =
+          completeSingleDisposalTriageAnswers.assetType === Residential || completeSingleDisposalTriageAnswers.assetType === NonResidential
+
+        val noOtherReliefs                 = reliefDetailsAnswers.otherReliefs.isEmpty
+
+        val underPreviousReturnLimit       = fillingOutReturn.previousSentReturns
+          .getOrElse(PreviousReturnData(summaries = List.empty, previousYearToDate = None))
+          .summaries
+          .length <= 10
+
+        val isSelf                         = completeSingleDisposalTriageAnswers.individualUserType.contains(Self)
+
+        val currentReturnIsEligible = eligibleAssetType && noOtherReliefs && underPreviousReturnLimit && isSelf
+
+        if (currentReturnIsEligible) {
+          val submissionIdsOfPreviousReturns = fillingOutReturn.previousSentReturns
+            .map(e => e.summaries)
+            .getOrElse(List.empty)
+            .map(e => e.submissionId)
+          val results = submissionIdsOfPreviousReturns.map {
+            returnsService
+              .displayReturn(fillingOutReturn.subscribedDetails.cgtReference, _)
+              .fold(
+                _ => false,
+                r =>
+                  r.completeReturn.fold(
```

Comment:
this would be clearer as a pattern match like:
```scala
r.completeReturn match {
  case s: CompleteSingleDisposalReturn => ...
  case _ => false
}
```